File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int n = s.length();
        int si = 0;
        int ei = 0;
        int count = 0;
        int len = 0;

        int[] fre = new int[128];

        while(ei < n){
            if(fre[s.charAt(ei)] == 1){
                count++;
            }
            fre[s.charAt(ei)]++;
            ei++;

            while(count > 0){
                if(fre[s.charAt(si)] == 2){
                    count--;
                }
                fre[s.charAt(si)]--;
                si++;
            }
            len = Math.max(len,ei-si);
        }
        return len;   
    }
}

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self,s):
        str1 = [False] * 256
        i = 0
        j = -1
        max_len = 0
        for i in range(len(s)):
            char_to_acquire = ord(s[i])
            while str1[char_to_acquire] == True:
                j += 1
                char_to_release = ord(s[j])
                str1[char_to_release] = False
            str1[char_to_acquire] = True
            ans_len = abs(i-j)
            max_len = max(ans_len,max_len)
            i = i + 1
        return max_len